### PR TITLE
New rule: use_to_and_as_if_applicable

### PR DESCRIPTION
**Pull request by JPaulsen**
Fixes https://github.com/dart-lang/linter/issues/580
Fixes https://github.com/dart-lang/linter/issues/581
And also is relevant to https://github.com/dart-lang/linter/issues/573

### CODE REVIEW

**Comment by JPaulsen** [comment]
0 lints in the SDK, if I delete the line that checks if already starts with to and as then apperas 7 lints:

```dart
/usr/local/google/home/casanueva/Development/sdk/sdk/lib/_internal/js_runtime/lib/js_array.dart 592:10 [lint] Start the name of the method with to/_to or as/_as if applicable
  Set<E> toSet() => new Set<E>.from(this);
         ^^^^^
/usr/local/google/home/casanueva/Development/sdk/sdk/lib/_internal/js_runtime/lib/js_array.dart 627:15 [lint] Start the name of the method with to/_to or as/_as if applicable
  Map<int, E> asMap() {
              ^^^^^
/usr/local/google/home/casanueva/Development/sdk/sdk/lib/convert/string_conversion.dart 186:22 [lint] Start the name of the method with to/_to or as/_as if applicable
  ClosableStringSink asStringSink() {
                     ^^^^^^^^^^^^
/usr/local/google/home/casanueva/Development/sdk/sdk/lib/async/future_impl.dart 273:13 [lint] Start the name of the method with to/_to or as/_as if applicable
  Stream<T> asStream() => new Stream<T>.fromFuture(this);
            ^^^^^^^^
/usr/local/google/home/casanueva/Development/sdk/sdk/lib/core/iterable.dart 359:10 [lint] Start the name of the method with to/_to or as/_as if applicable
  Set<E> toSet() => new Set<E>.from(this);
         ^^^^^
/usr/local/google/home/casanueva/Development/sdk/sdk/lib/collection/iterable.dart 91:10 [lint] Start the name of the method with to/_to or as/_as if applicable
  Set<E> toSet() => new Set<E>.from(this);
         ^^^^^
/usr/local/google/home/casanueva/Development/sdk/sdk/lib/collection/list.dart 330:15 [lint] Start the name of the method with to/_to or as/_as if applicable
  Map<int, E> asMap() {
              ^^^^^
```

We did not know if the lint was going to work well, but it seems good.

File: lib/src/rules/use_to_and_as_if_applicable.dart [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  Foo asFoo() {
+    return new Foo.from(this);
+  }
+}
+```
+
+''';
+
+bool _beginsWithAsOrTo(String name) {
+  final regExp = new RegExp(r"(to|as|_to|_as)[A-Z].*", caseSensitive: true);
```
Comment (bwilkerson) [inline lib/src/rules/use_to_and_as_if_applicable.dart]: Do we need the `.*`, given that we're matching the prefix?

**Comment by JPaulsen** [comment]
@bwilkerson I don't thinks so, we noticed that in general method that uses to and as, only return a new Something(this); but I do not know how to difference them

**Comment by JPaulsen** [comment]
Maybe we are not explaining the rule well, the idea is that if you have a method that does something like method() => new OtherClase(this); and it is not a getter then should be a toOtherClase or an asOtherClase method. If inside the method keeps the reference to the parameter and changes it when changing itself, then should be an as, but that is really hard to decide without false positives.

**Comment by bwilkerson** [comment]
Ah, I thought the rule was checking for two different cases but not differentiating in the error message. If it isn't, then a single message is appropriate.

File: lib/src/rules/use_to_and_as_if_applicable.dart [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        return _checkExpression(statement.expression);
+      }
+    }
+    return false;
+  }
+
+  bool _checkExpression(Expression expression) {
+    if (expression is ParenthesizedExpression) {
+      return _checkExpression(expression.expression);
+    }
```
Comment (bwilkerson) [inline lib/src/rules/use_to_and_as_if_applicable.dart]: I was reminded the other day that there's a better way to write this:
```
expression = expression.unParenthesized;
```
Several recent rules could be cleaned up.

File: lib/src/rules/use_to_and_as_if_applicable.dart (hunk 1 above)
Comment (JPaulsen) [inline lib/src/rules/use_to_and_as_if_applicable.dart]: Fixed

File: lib/src/rules/use_to_and_as_if_applicable.dart (hunk 2 above)
Comment (JPaulsen) [inline lib/src/rules/use_to_and_as_if_applicable.dart]: Fixed and issue added.

**Pull request merged**